Transition SimplifyPointerBitcastPass to opaque pointers

omarahmed1111 opened the pull request:
# PR changes
* **Modify `SimplifyPointerBitcastPass::runOnBitcastFromGEP`**: This function should simplify cases when we bitcast a pointer that is generated from gep instruction. Before opaque pointers that case was valid as we could transform a pointer type to another pointer type like (i64* to i32*). After opaque pointer changes that case can't happen and will always be a trivial bitcast as we don't have any pointer type other than "ptr". Trivial bitcasts will be handled with `runOnTrivialBitcast`. So, we should delete this function after full transformation to opaque pointers.  
Example: `runOnTrivialBitcast` will delete the bitcast instruction. 
```
%gep = getelementptr i32, ptr %ptr, i32 0     
%cast = bitcast ptr %gep to ptr
```

* **Modify `SimplifyPointerBitcastPass::runOnGEPFromBitcastCstExpr`**: This function should simplify cases when we have gep instruction with its pointer operand as a bitcast constant expression. Which will have a similar trivial case as `SimplifyPointerBitcastPass::runOnBitcastFromGEP`. So, we should delete this function after full transformation to opaque pointers. 
Example:  
`%gep = getelementptr i32, ptr bitcast (ptr %ptr to ptr), i32 0`  
LLVM passes will transform it to `getelementptr i32, ptr %ptr, i32 0`

* **Modify `SimplifyPointerBitcastPass::runOnGEPFromGEP**: This function handles cases when a gep instruction pointer operand is another gep instruction. Implicit casts could occur after opaque pointer transition. So, we make sure that there is no implicit cast before optimizing this cases.

* **Add some trivial and simple bitcast test cases**

### Related issues
https://github.com/google/clspv/issues/816

**This contribution is being made by Codeplay on behalf of Samsung.**

omarahmed1111 commented on test/PointerCasts/trivial_casts.ll:
```diff
@@ -0,0 +1,68 @@
+; CHECK:   ret void
+; CHECK: }
+
+define void @test4(i32* %in, float* %out) {
+entry:
+  %gep = getelementptr i32, i32* %in, i32 1     
+  %cast = bitcast i32* %gep to float*
+  %val = load float, float* %cast
+  store float %val, float* %out
+  ret void
+}
```
Not a trivial bitcast but thought adding it as a prove of concept for opaque pointer transformation. as `opaque_trivial_casts.ll` have its similar version but with opaque pointer (which will be trivial). It also should cover a case that is not currently tested for `runOnBitcastFromGEP` (this function should be deleted after full transformation anyway).

alan-baker reviewed (changes requested):
Mostly there. I think the GEP from GEP case needs additional checking now in an opaque pointer world though. There might be an implicit pointer cast between them so I think we need a check that the result element type of the first GEP is the same as the source element type of the second GEP. I believe that also works with transparent pointers too. It would be good to make sure there is coverage of those cases for opaque pointers too in the tests.

omarahmed1111 commented:
> Mostly there. I think the GEP from GEP case needs additional checking now in an opaque pointer world though. There might be an implicit pointer cast between them so I think we need a check that the result element type of the first GEP is the same as the source element type of the second GEP. I believe that also works with transparent pointers too. It would be good to make sure there is coverage of those cases for opaque pointers too in the tests.

Fixed and added small simple tests for that.
